extend kep-19 for single semver (#1103)
kudobuilder/kudo: keps/0019-package-api-versioning.md

### Operator

> An opinionated way to run a workload of a specific domain built on KUDO.

Reviewer: Let's keep the more general old description here.

- Visually simple
- It can be known just from the package version whether or not it contains a certain operator-related feature or bug fix

> - Version is a combination of the application and operator in a single SemVer.

Reviewer: IMO, it is not :). It can be anything SemVer, we're not forcing operator devs to anything here. Though, with things like backports in mind, it makes sense to couple certain application version changes to operator version changed.

Reviewer 2: I think getting a precise phrasing here is tricky, but afaik the intention is to mean that the *version* is tied to an *application version* plus a specific set of operator-features. (@zmalik to confirm 😛)

Author: yes, I meant originally from where it's derived

> - Version is a derived from the base tech and operator development as a single semver.

Reviewer 2: ```suggestion
- Version is a combination of the application and operator in a single SemVer.
```

  Example:

  | Time | Version  | App version        | Comment                                                     |
  | ---- | -------- | ------------------ | ----------------------------------------------------------- |
  | T0   | 1.0.0    | 2.3.0              | Initial release based on app version 2.3.x                  |
  | T1   | 2.0.0    | 3.0.0              | Release based on app version 3.0.x                          |
  | T2   | 2.1.0    | 3.0.0              | New operator-related feature A for latest app version 3.0.x |
  | T2   | 1.1.0    | 2.3.0              | Back-port feature A for app version 2.3.x on demand         |
  | T3   | 1.1.1    | 2.3.1              | Bug fix release for app version 2.3.x                       |
  | T4   | 2.2.0    | 3.0.0              | Important bug fix B in operator code for latest 3.0.x       |
  | T4   | 1.2.0    | 2.3.1              | Back-port bug fix B in operator code for 2.3.x on demand    |
  | T5   | 2.2.1    | 3.0.0              | Small bug fix C in operator code for latest 3.0.x           |

>   | T5   | 1.2.1    | 2.3.1              | Back-port bug fix C in operator code for 2.3.x on demand    |

Reviewer 2: What happens when a breaking change needs to happen the `1.x` version family? Does it bump to the next available major version, which in this simplified scenario would be `3.x`?

Reviewer 2: It looks like the part of the version that keeps track of what is in the operator wrt operator-related features is being manually synced between version families? e.g., `1.2.1` has the same operator-related features and bug fixes as `2.2.1` because they're both `x.2.1`?

Author: when we had done GA an operator are we allowed to introduce breaking changes? It will need a reset of major version ofc. And versioning will be just one thing, as we would definitively need to provide an upgrade path for people already running a "GA" version. 
And it will be always done with supported versions in mind. 

In case we just reset Kafka from 0. As its a breaking change. We support two versions, so we will have to introduce two major versions, explaining the upgrade path to them also.

##### Cons

- Visually more complex
- Makes it complex to know which operator-related features are available in which versions.  Operator revision for different app versions are unrelated

> - Upgrades paths are unclear. Upgrading from `3.11.4-0.1.2` to `3.12.0-0.1.2` might result in missing some features that were present in `3.11.4-0.1.2`

Reviewer 2: > Upgrades paths are unclear

The same is true for single version afaict. The rest of the sentence is fair.

In practice, in the example given above it would be more likely that `3.12.0-0.1.2` would have more features then `3.11.4-0.1.2`, even though it will still be theoretically possible that the later app version doesn't have a feature from an earlier app version.

Author: this is not the same for single semver. Where the development is continuous. Here with operator revision reset. It's hard to predict if `3.11.4-0.1.2` to `3.12.0-0.1.2` is just base tech bump or features bump. Or if some features are introduced or same feature parity will be in `3.12.0-0.1.3` or `3.12.0-0.1.1`

Reviewer 2: Yep, that's what I meant with "The rest of the sentence is fair." The "Upgrades paths are unclear" part is also for the single version afaict, just in a different way.

Reviewer 2: Is this suggestion fair? https://github.com/kudobuilder/kudo/pull/1103#discussion_r350165019

Author: yeah, I will add that,  I didn't have given enough time to go through all cons, as most of the cons of a single version also apply for the composite version.
`Operator users care about the application version, not a "synthetic" version` can be translated to `Operator users care about the operator features also` 
And `it might be possible that subsequent package versions with different app versions don't contain an operator-related feature or bug fix that only makes sense for a specific app version` is true for composite version also. `3.11.4-0.1.2` might have a feature that `3.12.0-0.1.2` don't have as that was specific to `3.11.4`

> - Upgrades paths are unclear. Upgrading from `3.11.4-0.1.2` to `3.12.0-0.1.2` might result in missing some features that were present in `3.11.4-0.1.2`

Reviewer 2: ```suggestion
- Upgrading from `3.11.4-0.1.2` to `3.12.0-0.1.2` might result in missing some operator-related features that were present in `3.11.4-0.1.2`, even if unlikely
```

Installs `kafka-3.0.0-1.1.0`.

**for single version:**

```
kudo install kafka --app-version 2.3.0
```

Installs `kafka-1.1.0`.

```
kudo install kafka --app-version 3.0.0
```

Installs `kafka-2.2.1`.

**Install a Kafka operator with a specific version**

Let's say a user wants to use Apache Kafka `3.0.0` and even though there's a `3.0.0` with a `1.1.0` revision released, they want the operator with revision `1.0.0` due to reasons like:
- revision `1.1.0` introduced a bug for which there's still no released fixes
- revision `1.1.0` changed in a way that will require them to invest time and resources to adapt

**for composite version:**

```
kudo install kafka --version 3.0.0-1.0.0
```

**for single version:**

it should be enough with single version
```

> kudo install kafka --version 2.2.0

Reviewer 3: so is this reversing the order of which version has primacy? Or is this getting the latest version of Kafka and installing it at that operator version?

Author: it separates `--version` from `--app-version`
